File: CrackingTheCodingInterview/Chapter1/one-six.py
# string compression

def check(string):
    new_length = length_check(string)*2
    if (len(string) <= new_length):
        return string
    return compress(string, new_length)

def length_check(string):
    length = 1
    for i in range(0, len(string)-1):
        if (string[i] != string[i+1]):
            length += 1
    return length

def compress(string, length):
    array = [None] * length
    array_index = 0
    count = 1
    for i in range(0, len(string)):
        if(i+1 >= len(string) or string[i] != string[i+1]):
            array[array_index] = string[i]
            array[array_index+1] = str(count)
            array_index += 2
            count = 1      
        else:
            count += 1
    return "".join(array)

File: CrackingTheCodingInterview/Chapter1/OneSix.java
package CrackingTheCodingInterview.Chapter1;

// String compression
public class OneSix {

    public static void main(String[] args)
    {

    }

    static String join(String str)
    {
        int length = checkLength(str);
        if (str.length() <= length)
        {
            return str;
        }
        return compress(str, length);
    }

    static int checkLength(String str)
    {
        int length = 1;
        for (int i = 0; i < str.length()-1; i++)
        {
            if (str.charAt(i) != str.charAt(i+1))
            {
                length++;
            }
        }
        return length;
    }

    static String compress(String str, int length)
    {
        StringBuilder array = new StringBuilder(length);
        // String[] array = new String[length];
        int count = 1;
        for (int i = 0; i < str.length(); i++)
        {
            if(i+1 >= str.length() || str.charAt(i) != str.charAt(i+1))
            {
                array.append(str.substring(i, i+1));
                array.append(count);
                count = 1;
            }
            else
            {
                count++;
            }
        }
        return array.toString();
    }
    
}